File: tron-backend/src/adapters/messenger.js
const axios = require('axios');
const { supabase } = require('../config/supabase');

/**
 * Broadcasts the AI Executive Summary to the team's communication channel
 */
async function broadcastSummary(communicationConfig, prTitle, prUrl, report, orgId) { 
    if (!communicationConfig) return;

    const { provider, webhook_url, channel_id, bot_token } = communicationConfig;

    try {
        if (provider === 'discord') {
            if (!webhook_url) throw new Error("Missing webhook_url");
            await sendDiscord(webhook_url, prTitle, prUrl, report);
            
        } else if (provider === 'discord_bot') {
            if (!channel_id) throw new Error("Missing channel_id");
            if (!orgId) throw new Error("Missing orgId to fetch Discord token");

            // Fetch the bot token from the DB dynamically based on the Organization
            const { data: integration, error } = await supabase
                .from('integrations')
                .select('token, secret_id')
                .eq('org_id', orgId)
                .in('provider', ['discord', 'discord_bot'])
                .limit(1)
                .single();

            let actualBotToken = integration?.token || bot_token;

            // Secure Vault Fallback
            if (integration?.secret_id && !integration?.token) {
                const { data: decryptedJson } = await supabase.rpc('get_decrypted_secret', {
                    p_secret_id: integration.secret_id
                });
                if (decryptedJson) {
                    const creds = JSON.parse(decryptedJson);
                    actualBotToken = creds.botToken || creds.bot_token || creds.token || actualBotToken;
                }
            }

            if (!actualBotToken) throw new Error("Missing bot_token in database for this organization.");
            
            await sendDiscordBot(actualBotToken, channel_id, prTitle, prUrl, report);
            
        } else if (provider === 'slack') {
            if (!webhook_url) throw new Error("Missing webhook_url");
            await sendSlack(webhook_url, prTitle, prUrl, report);
            
        } else {
            console.warn(`⚠️  [Messenger] Unsupported communication provider: '${provider}'`);
        }
    } catch (error) {
        console.error(`❌ [Messenger] Failed to broadcast to ${provider}:`, error.message);
    }
}

// ==========================================
// 🌟 The Discord Bot API Route
// ==========================================
async function sendDiscordBot(botToken, channelId, prTitle, prUrl, report) {
    const url = `https://discord.com/api/v10/channels/${channelId}/messages`;

    const payload = {
        embeds: [{
            title: `🤖 T.R.O.N. Intel: ${prTitle}`,
            url: prUrl,
            color: 3447003, // TRON Blue
            fields: [
                { name: '🏷️ Category', value: report.intent || 'Unknown' },
                { name: '📝 Summary', value: report.executive_summary || 'No summary provided.' },
                { name: '🚀 Business Impact', value: report.business_impact || 'No impact analysis provided.' }
            ],
            footer: { text: 'TRON V3 AI Pipeline' },
            timestamp: new Date().toISOString()
        }]
    };

    await axios.post(url, payload, {
        headers: {
            'Authorization': `Bot ${botToken}`, 
            'Content-Type': 'application/json'
        }
    });
    
    console.log('✅ [Messenger] Successfully broadcasted AI Intel via Custom Discord Bot.');
}

// ==========================================
// LEGACY WEBHOOK ROUTES
// ==========================================
async function sendDiscord(webhookUrl, prTitle, prUrl, report) {
    const payload = {
        embeds: [{
            title: `🤖 T.R.O.N. Intel: ${prTitle}`,
            url: prUrl,
            color: 3447003, 
            fields: [
                { name: '🏷️ Category', value: report.intent || 'Unknown' },
                { name: '📝 Summary', value: report.executive_summary || 'No summary provided.' },
                { name: '🚀 Business Impact', value: report.business_impact || 'No impact analysis provided.' }
            ],
            footer: { text: 'TRON V3 AI Pipeline' },
            timestamp: new Date().toISOString()
        }]
    };

    await axios.post(webhookUrl, payload);
    console.log('✅ [Messenger] Successfully broadcasted AI Intel to Discord.');
}

async function sendSlack(webhookUrl, prTitle, prUrl, report) {
    const payload = {
        blocks: [
            {
                type: "header",
                text: { type: "plain_text", text: `🤖 T.R.O.N. Intel: ${prTitle}` }
            },
            {
                type: "section",
                text: { type: "mrkdwn", text: `*<${prUrl}|View Pull Request>*` }
            },
            {
                type: "section",
                fields: [
                    { type: "mrkdwn", text: `*🏷️ Category:*\n${report.intent || 'Unknown'}` },
                    { type: "mrkdwn", text: `*📝 Summary:*\n${report.executive_summary || 'No summary'}` },
                    { type: "mrkdwn", text: `*🚀 Business Impact:*\n${report.business_impact || 'No impact'}` }
                ]
            }
        ]
    };

    await axios.post(webhookUrl, payload);
    console.log('✅ [Messenger] Successfully broadcasted AI Intel to Slack.');
}

module.exports = {
    broadcastSummary
};

File: tron-go-backend/internal/adapters/messenger.go
package adapters

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/tron-v3.1/tron-go-backend/internal/models"
	"github.com/tron-v3.1/tron-go-backend/internal/services"
	"github.com/tron-v3.1/tron-go-backend/pkg/database"
)

// CommunicationConfig maps the JSON settings from the database
type CommunicationConfig struct {
	Provider   string `json:"provider"`
	WebhookURL string `json:"webhook_url"`
	ChannelID  string `json:"channel_id"`
	BotToken   string `json:"bot_token"`
}

// AIReport standardizes the AI output so the messenger doesn't care which LLM generated it
type AIReport struct {
	Intent           string
	ExecutiveSummary string
	BusinessImpact   string
}

// MessengerAdapter handles broadcasting to team communication channels
type MessengerAdapter struct {
	HTTPClient *http.Client
}

// NewMessengerAdapter initializes the client with a standard timeout
func NewMessengerAdapter() *MessengerAdapter {
	return &MessengerAdapter{
		HTTPClient: &http.Client{Timeout: 10 * time.Second},
	}
}

// BroadcastSummary routes the message to the correct platform
func (api *MessengerAdapter) BroadcastSummary(config CommunicationConfig, prTitle, prURL string, report AIReport, orgID string) {
	if config.Provider == "" {
		return
	}

	switch config.Provider {
	case "discord":
		if config.WebhookURL == "" {
			fmt.Println("❌ [Messenger] Missing webhook_url for Discord")
			return
		}
		api.sendDiscord(config.WebhookURL, prTitle, prURL, report)

	case "discord_bot":
		if config.ChannelID == "" || orgID == "" {
			fmt.Println("❌ [Messenger] Missing channel_id or orgId for Discord Bot")
			return
		}
		api.sendDiscordBot(config, prTitle, prURL, report, orgID)

	case "slack":
		if config.WebhookURL == "" {
			fmt.Println("❌ [Messenger] Missing webhook_url for Slack")
			return
		}
		api.sendSlack(config.WebhookURL, prTitle, prURL, report)

	default:
		fmt.Printf("⚠️ [Messenger] Unsupported communication provider: '%s'\n", config.Provider)
	}
}

// ==========================================
// 🌟 The Discord Bot API Route (Smart Vault Fallback)
// ==========================================
func (api *MessengerAdapter) sendDiscordBot(config CommunicationConfig, prTitle, prURL string, report AIReport, orgID string) {
	var integration models.Integration
	actualBotToken := config.BotToken

	// 1. Fetch the integration record from the database
	err := database.DB.Where("org_id = ? AND provider IN ?", orgID, []string{"discord", "discord_bot"}).First(&integration).Error

	// 2. 🌟 THE FIX: Smart Vault Fallback in Go
	if err == nil && integration.SecretID != nil && actualBotToken == "" {
		decryptedSecret, decErr := services.GetDecryptedSecret(*integration.SecretID)

		if decErr == nil && decryptedSecret != "" {
			// Try to parse it as JSON first
			var creds map[string]interface{}
			if parseErr := json.Unmarshal([]byte(decryptedSecret), &creds); parseErr == nil {
				if val, ok := creds["botToken"].(string); ok {
					actualBotToken = val
				} else if val, ok := creds["bot_token"].(string); ok {
					actualBotToken = val
				} else if val, ok := creds["token"].(string); ok {
					actualBotToken = val
				}
			} else {
				// If JSON parsing fails, it means the secret IS the raw token!
				actualBotToken = decryptedSecret
			}
		}
	}

	if actualBotToken == "" {
		fmt.Println("❌ [Messenger] Missing bot_token in database for this organization.")
		return
	}

	url := fmt.Sprintf("https://discord.com/api/v10/channels/%s/messages", config.ChannelID)
	payload := api.buildDiscordPayload(prTitle, prURL, report)
	payloadBytes, _ := json.Marshal(payload)

	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
	req.Header.Set("Authorization", "Bot "+actualBotToken)
	req.Header.Set("Content-Type", "application/json")

	resp, reqErr := api.HTTPClient.Do(req)
	if reqErr != nil || resp.StatusCode >= 400 {
		fmt.Printf("❌ [Messenger] Failed to send Discord Bot message. Status: %d\n", resp.StatusCode)
		return
	}
	defer resp.Body.Close()

	fmt.Println("✅ [Messenger] Successfully broadcasted AI Intel via Custom Discord Bot.")
}

// ==========================================
// LEGACY WEBHOOK ROUTES
// ==========================================
func (api *MessengerAdapter) sendDiscord(webhookURL, prTitle, prURL string, report AIReport) {
	payload := api.buildDiscordPayload(prTitle, prURL, report)
	payloadBytes, _ := json.Marshal(payload)

	resp, err := api.HTTPClient.Post(webhookURL, "application/json", bytes.NewBuffer(payloadBytes))
	if err != nil || resp.StatusCode >= 400 {
		fmt.Println("❌ [Messenger] Failed to send Discord Webhook message.")
		return
	}
	defer resp.Body.Close()

	fmt.Println("✅ [Messenger] Successfully broadcasted AI Intel to Discord.")
}

func (api *MessengerAdapter) sendSlack(webhookURL, prTitle, prURL string, report AIReport) {
	// Build Slack Blocks Payload
	payload := map[string]interface{}{
		"blocks": []map[string]interface{}{
			{
				"type": "header",
				"text": map[string]string{"type": "plain_text", "text": "🤖 T.R.O.N. Intel: " + prTitle},
			},
			{
				"type": "section",
				"text": map[string]string{"type": "mrkdwn", "text": fmt.Sprintf("*<%s|View Pull Request>*", prURL)},
			},
			{
				"type": "section",
				"fields": []map[string]string{
					{"type": "mrkdwn", "text": "*🏷️ Category:*\n" + fallbackText(report.Intent, "Unknown")},
					{"type": "mrkdwn", "text": "*📝 Summary:*\n" + fallbackText(report.ExecutiveSummary, "No summary")},
					{"type": "mrkdwn", "text": "*🚀 Business Impact:*\n" + fallbackText(report.BusinessImpact, "No impact")},
				},
			},
		},
	}

	payloadBytes, _ := json.Marshal(payload)

	resp, err := api.HTTPClient.Post(webhookURL, "application/json", bytes.NewBuffer(payloadBytes))
	if err != nil || resp.StatusCode >= 400 {
		fmt.Println("❌ [Messenger] Failed to send Slack Webhook message.")
		return
	}
	defer resp.Body.Close()

	fmt.Println("✅ [Messenger] Successfully broadcasted AI Intel to Slack.")
}

// ==========================================
// PRIVATE HELPERS
// ==========================================

// buildDiscordPayload standardizes the Embed look across Webhooks and Bots
func (api *MessengerAdapter) buildDiscordPayload(prTitle, prURL string, report AIReport) map[string]interface{} {
	return map[string]interface{}{
		"embeds": []map[string]interface{}{
			{
				"title":     "🤖 T.R.O.N. Intel: " + prTitle,
				"url":       prURL,
				"color":     3447003,
				"timestamp": time.Now().UTC().Format(time.RFC3339),
				"footer":    map[string]string{"text": "TRON V3 AI Pipeline"},
				"fields": []map[string]string{
					{"name": "🏷️ Category", "value": fallbackText(report.Intent, "Unknown")},
					{"name": "📝 Summary", "value": fallbackText(report.ExecutiveSummary, "No summary provided.")},
					{"name": "🚀 Business Impact", "value": fallbackText(report.BusinessImpact, "No impact analysis provided.")},
				},
			},
		},
	}
}

// fallbackText ensures we don't send empty fields which cause Discord/Slack APIs to reject the payload
func fallbackText(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
